nighthawk/source/client/stream_decoder.cc

```diff
@@ -0,0 +1,55 @@
+#include "nighthawk/source/client/stream_decoder.h"
+
+#include "common/http/http1/codec_impl.h"
+#include "common/http/utility.h"
+
+namespace Nighthawk {
+namespace Client {
+
+void StreamDecoder::decodeHeaders(Envoy::Http::HeaderMapPtr&& headers, bool end_stream) {
+  ASSERT(!complete_);
+  complete_ = end_stream;
+  response_headers_ = std::move(headers);
+  if (complete_) {
+    onComplete(true);
+  }
+}
+
+void StreamDecoder::decodeData(Envoy::Buffer::Instance&, bool end_stream) {
+  ASSERT(!complete_);
+  complete_ = end_stream;
+  if (complete_) {
+    onComplete(true);
+  }
+}
+
+void StreamDecoder::decodeTrailers(Envoy::Http::HeaderMapPtr&&) { NOT_IMPLEMENTED_GCOVR_EXCL_LINE; }
+
+void StreamDecoder::onComplete(bool success) {
+  if (success && measure_latencies_) {
+    latency_statistic_.addValue((time_source_.monotonicTime() - request_start_).count());
+  }
+  ASSERT(complete_);
+  decoder_completion_callback_.onComplete(success, *response_headers_);
+  caller_completion_callback_();
+  delete this;
```

Comment:
FWIW, I've found historically self-deleting to be a dangerous pattern when combined with polymorphism, since any subclass needs to ensure it doesn't have any code after the `delete`. Did you consider using Envoy's `deferredDelete`? This is a much safer RCU-like approach. If you want to stick with self-deleting, I recommend making the class `final`.

Reply:
Nice, done